### PR TITLE
Split block and evict task

### PR DESCRIPTION
**Pull request by jaewan**
## Why are these changes needed?
Previous version implemented blocktasks but with currently submitted task which caused a deadlock (NSDI23/single~/debug/deadlock.py)
This PR is to find the lowest priority from object_table and set blocktask according to it.
+ add separation from blocking new tasks being dispatched only and block spill (enableBlockTaskSpill)
with blocktasks with non-spill, there are 2 deadlock cases found in NSDI23/single~/debug/deadlock1.py and 2.py
All files from NSDI23 directory are microbenchmark. You can ignore them.
Many changes here are from ClangFormat changes. I listed the files you can ignore

# Files changes for BlockTasks
- object_manager/plasma/create_request_queue.cc: L185 ~ 218
   L34 ~ 167 are just ClangFormat and argument added for block_task threshold (which not used for now) so you can ignore this
- raylet/node_manager.cc : callback function for blocktasks

# Files changes for searching the lowest priority object in the object_table
- object_manager/plasma/common.h : Added GetPriority to find priority per object
- object_manager/plasma/create_request_queue.h: getting the lowest priority from object_table to set block_task
- object_manager/plasma/object_lifecycle_manager.cc
- object_manager/plasma/object_lifecycle_manager.h
- object_manager/plasma/object_store.cc: Search for the lowest pri obj from obejct_table
- object_manager/plasma/object_store.h
- object_manager/plasma/protocol.cc: copying priority

# Files you can skip 
Many changes in arguments are from ClangFormat. You can ignore these changes from the following files
- core_worker/core_worker.cc
- object_manager/plasma/create_request_queue.cc: L34 ~ 167 (These changes are ClangFormat and for block_task threshold)
    You want to check out L185 ~ 218
- object_manager/plasma/eviction_policy.cc : some debug messages
- object_manager/plasma/object_store.cc: check object store if it reached a threshold (for threshold blocktask)
- object_manager/plasma/store_runner.cc
- raylet/local_object_manager.cc
- raylet/local_object_manager.h
- raylet/scheduling/cluster_task_manager.cc: ClangFormat and some argument added for blocktask callback function
- raylet/test/local_object_manager_test.cc

### CODE REVIEW

File: src/ray/common/ray_config_def.h [hunk 1]
```diff
@@ -374,7 +374,20 @@ RAY_CONFIG(int64_t, min_spilling_size, 100 * 1024 * 1024)
...
+RAY_CONFIG(float, evict_tasks_threshold, 1.0)
+
+// Whether to use BlockTasks 
+RAY_CONFIG(bool, enable_BlockTasks, true)
+
+// Whether to use EvictTasks when spill required
+RAY_CONFIG(bool, enable_EvictTasks, false)
+
+// Whether to use EvictTasks when spill required
+RAY_CONFIG(bool, enable_BlockandEvictTasks, false)
```
Comment (stephanie-wang) [inline src/ray/common/ray_config_def.h]: Can you remove this? I think it's confusing and unnecessary with the other two flags.

File: src/ray/object_manager/plasma/create_request_queue.cc [hunk 2]
```diff
@@ -173,12 +196,31 @@ Status CreateRequestQueue::ProcessRequests() {
...
-      }
+	  if(RayConfig::instance().enable_BlockandEvictTasks()){
+		on_object_creation_blocked_callback_(queue_it->first.first);
+		on_object_evict_callback_(queue_it->first.first);
+        if (!should_spill_) {
+          RAY_LOG(INFO) << "Object creation of priority " << queue_it->first.first << " blocked";
+          return Status::TransientObjectStoreFull("Waiting for higher priority tasks to finish");
+        }
+		RAY_LOG(INFO) << "[JAE_DEBUG] should_spill set";
+			    SetShouldSpill(false);
```
Comment (stephanie-wang) [inline src/ray/object_manager/plasma/create_request_queue.cc]: I don't think we should set should_spill_ here, right? We should allow the callbacks to set this later on. What was the reason for setting it?

File: src/ray/object_manager/plasma/create_request_queue.cc [hunk 3]
```diff
@@ -214,7 +256,9 @@ Status CreateRequestQueue::ProcessRequests() {
 
   // If we make it here, then there is nothing left in the queue. It's safe to
   // run new tasks again.
-  RAY_UNUSED(on_object_creation_blocked_callback_(ray::Priority()));
+  if(!RayConfig::instance().enable_BlockandEvictTasks() || RayConfig::instance().enable_BlockTasks()){
```
Comment (stephanie-wang) [inline src/ray/object_manager/plasma/create_request_queue.cc]: Not sure if this is right. Shouldn't we call the callback when we are back under the threshold, inside the main loop?

File: src/ray/object_manager/plasma/create_request_queue.cc [hunk 4]
```diff
@@ -173,12 +196,31 @@ Status CreateRequestQueue::ProcessRequests() {
...
       }
 
-      // Notify the scheduler that object creation is blocked at this priority.
-      on_object_creation_blocked_callback_(queue_it->first.first);
-      if (!should_spill_) {
-        RAY_LOG(INFO) << "Object creation of priority " << queue_it->first.first << " blocked";
-        return Status::TransientObjectStoreFull("Waiting for higher priority tasks to finish");
-      }
+	  if(RayConfig::instance().enable_BlockandEvictTasks()){
+		on_object_creation_blocked_callback_(queue_it->first.first);
```
Comment (stephanie-wang) [inline src/ray/object_manager/plasma/create_request_queue.cc]: Instead of having two separate callbacks, can we do one callback that sets the blocked priority and a flag for whether to also evict tasks?

I think this will be cleaner and also easier to debug. Since the callbacks here are concurrent, it's going to be hard to figure out what's going on if there are too many callbacks.

**Review by stephanie-wang — COMMENTED** [review]
Can you separate the threshold changes into a separate PR, so this PR is just the change to add an option for blocking/evicting tasks? We can sync in person on how to implement the threshold.

File: src/ray/object_manager/plasma/create_request_queue.cc (hunk 3 above)
Comment (stephanie-wang) [reply src/ray/object_manager/plasma/create_request_queue.cc]: 1. ProcessRequest() will set if we are under or over the threshold.
2. If we are over the threshold, we should call BlockTasks() on the lower priority object currently in the object store.
3. If we are under the threshold, then we should call BlockTasks(Priority()).

This means we also need to track what the lowest priority object currently is.

File: src/ray/object_manager/plasma/create_request_queue.cc [hunk 5]
```diff
@@ -173,8 +196,18 @@ Status CreateRequestQueue::ProcessRequests() {
         oom_start_time_ns_ = now;
       }
 
-      // Notify the scheduler that object creation is blocked at this priority.
-      on_object_creation_blocked_callback_(queue_it->first.first);
+      if(RayConfig::instance().enable_BlockTasks()){
+        RAY_LOG(DEBUG) << "[JAE_DEBUG] calling object_creation_blocked_callback priority "
+	    << queue_it->first.first.score;
+	    on_object_creation_blocked_callback_(queue_it->first.first , true, false);
```
Comment (stephanie-wang) [inline src/ray/object_manager/plasma/create_request_queue.cc]: We should call the callback a single time with the right flags, instead of calling it twice (this is to avoid having concurrent callbacks).

File: src/ray/object_manager/plasma/store.cc [hunk 6]
```diff
@@ -161,25 +167,28 @@ PlasmaError PlasmaStore::HandleCreateObjectRequest(const std::shared_ptr<Client>
                    << ", metadata_size=" << object_info.metadata_size;
   }
   //TODO(Jae) Erase this later
-  /*
   const int64_t footprint_limit = allocator_.GetFootprintLimit();
-  const float allocated_percentage =
-        static_cast<float>(allocator_.Allocated()) / footprint_limit;
-  if(allocated_percentage >= block_tasks_threshold_){
-	  blockTasks();
+  float allocated_percentage;
```
Comment (stephanie-wang) [inline src/ray/object_manager/plasma/store.cc:166]: I thought we were going to do the thresholding in a separate PR? In this PR, I think we can always assume that we should block/evict tasks if object creation is blocked by OOM.

File: src/ray/object_manager/plasma/create_request_queue.cc [hunk 7]
```diff
@@ -173,11 +182,40 @@ Status CreateRequestQueue::ProcessRequests() {
         oom_start_time_ns_ = now;
       }
 
-      // Notify the scheduler that object creation is blocked at this priority.
-      on_object_creation_blocked_callback_(queue_it->first.first);
-      if (!should_spill_) {
-        RAY_LOG(INFO) << "Object creation of priority " << queue_it->first.first << " blocked";
-        return Status::TransientObjectStoreFull("Waiting for higher priority tasks to finish");
+      if (RayConfig::instance().enable_BlockTasks()) {
```
Comment (stephanie-wang) [inline src/ray/object_manager/plasma/create_request_queue.cc]: Shouldn't this also check if `block_tasks_required` is true?

File: src/ray/object_manager/plasma/create_request_queue.cc [hunk 8]
```diff
@@ -173,11 +182,40 @@ Status CreateRequestQueue::ProcessRequests() {
...
+        RAY_LOG(DEBUG) << "[JAE_DEBUG] calling object_creation_blocked_callback priority "
+                       << lowest_pri;
+        on_object_creation_blocked_callback_(lowest_pri, true, false);
+		if(!RayConfig::instance().enable_BlockTasksSpill()){
+		  spill_objects_callback_();
```
Comment (stephanie-wang) [inline src/ray/object_manager/plasma/create_request_queue.cc]: Could we do something simpler like just return early here?

File: src/ray/object_manager/plasma/create_request_queue.cc [hunk 9]
```diff
@@ -214,15 +252,19 @@ Status CreateRequestQueue::ProcessRequests() {
 
   // If we make it here, then there is nothing left in the queue. It's safe to
   // run new tasks again.
-  RAY_UNUSED(on_object_creation_blocked_callback_(ray::Priority()));
+  if (RayConfig::instance().enable_BlockTasks() && 
+		  !RayConfig::instance().enable_EvictTasks()) {
```
Comment (stephanie-wang) [inline src/ray/object_manager/plasma/create_request_queue.cc]: Why do we check enableEvictTasks here?

File: src/ray/object_manager/plasma/create_request_queue.cc (hunk 7 above)
Comment (stephanie-wang) [reply src/ray/object_manager/plasma/create_request_queue.cc]: Also, where is the enable_EvictTasks() flag used? I only see it at the end of the loop but I thought it should also use it here?

File: src/ray/object_manager/plasma/object_store.cc [hunk 10]
```diff
@@ -51,6 +51,21 @@ const LocalObject *ObjectStore::CreateObject(const ray::ObjectInfo &object_info,
   return entry;
 }
 
+ray::Priority ObjectStore::GetLowestPriObject() {
+  // Return the lowest priority object in object_table
+  auto it = object_table_.begin();
+  ray::Priority lowest_priority = it->second->GetPriority();
```
Comment (stephanie-wang) [inline src/ray/object_manager/plasma/object_store.cc:59]: Won't this segfault if the object table is empty?

File: src/ray/object_manager/plasma/object_store.cc [hunk 11]
```diff
@@ -51,6 +51,21 @@ const LocalObject *ObjectStore::CreateObject(const ray::ObjectInfo &object_info,
...
+  it++;
+  for (; it != object_table_.end(); it++){
+	ray::Priority p = it->second->GetPriority();
+    if(lowest_priority < p){
```
Comment (stephanie-wang) [inline src/ray/object_manager/plasma/object_store.cc:62]: Isn't this actually choosing the highest priority object instead of the lowest?

File: src/ray/object_manager/plasma/store.cc [hunk 12]
```diff
@@ -160,26 +162,38 @@ PlasmaError PlasmaStore::HandleCreateObjectRequest(const std::shared_ptr<Client>
...
+
+  if (footprint_limit != 0) {
+    allocated_percentage = static_cast<float>(allocator_.Allocated()) / footprint_limit;
+  } else {
+    allocated_percentage = 0;
+  }
+  if (block_tasks_required != nullptr) {
+    if (allocated_percentage >= block_tasks_threshold_) {
```
Comment (stephanie-wang) [inline src/ray/object_manager/plasma/store.cc:174]: How is the block tasks threshold set?

Comment (jaewan) [reply src/ray/object_manager/plasma/store.cc:174]: It is set from ray_config. In default they are 1.0

File: src/ray/object_manager/plasma/create_request_queue.cc [hunk 13]
```diff
@@ -168,6 +168,14 @@ Status CreateRequestQueue::ProcessRequests() {
     if (spilling_required) {
       spill_objects_callback_();
     }
+	//Block and evict tasks are called if the object store reaches over a threshold
+	if(RayConfig::instance().enable_BlockTasks() && block_tasks_required){
+        on_object_creation_blocked_callback_(lowest_pri, true, false);
```
Comment (stephanie-wang) [inline src/ray/object_manager/plasma/create_request_queue.cc]: I think we may have talked about this earlier, but can we change this so that we only call the callback once? This is to reduce the chance of a race condition since these callbacks are async.

**Review by stephanie-wang — APPROVED** [review]
